src/union_find.ml

```diff
@@ -44,77 +44,84 @@ end
   deepest node takes Theta(N) time.  With the balancing scheme of never
   increasing the rank of a node unnecessarily, it would take O(log N).
 *)
-type 'a root = {
-  mutable value: 'a;
-  mutable rank: int;
-}
 
-type 'a t = { mutable node : 'a node; }
+(* NOTE: This does not work yet due to an OCaml 4.04 beta bug.  Should be
+   fixed before the release of OCaml 4.04. *)
+(* type ('a, 'kind) tree = *)
+(*   | Root : { mutable value : 'a; mutable rank : int } -> ('a, [ `root ]) tree *)
+(*   | Inner : { mutable parent : 'a node } -> ('a, [ `inner ]) tree *)
+(*  *)
+(* and 'a node = Node : ('a, _) tree -> 'a node  [@@ocaml.unboxed] *)
+(*  *)
+(* type 'a t = ('a, [ `inner ]) tree *)
 
-and 'a node =
-  | Inner of 'a t (* [Inner x] is a node whose parent is [x]. *)
-  | Root of 'a root
+type ('a, 'kind, 'parent) tree =
+  | Root : { mutable value : 'a; mutable rank : int } ->
+    ('a, [ `root ], 'parent) tree
+  | Inner : { mutable parent : 'parent } -> ('a, [ `inner ], 'parent) tree
+
+type 'a node = Node : ('a, _, 'a node) tree -> 'a node  [@@ocaml.unboxed]
+
+type 'a t = ('a, [ `inner ], 'a node) tree
 
 let invariant t =
-  let rec loop t depth =
-    match t.node with
-    | Inner t -> loop t (depth + 1)
-    | Root r -> assert (depth <= r.rank)
+  let rec loop (Inner inner) depth =
+    match inner.parent with
+    | Node (Inner _ as parent) -> loop parent (depth + 1)
+    | Node (Root r) -> assert (depth <= r.rank)
   in
   loop t 0
 
-let create v = { node = Root { value = v; rank = 0; }; }
+let create v = Inner { parent = Node (Root { value = v; rank = 0 }) }
 
-(* invariants:
-   [inner.node] = [inner_node] = [Inner t].
-   [descendants] are the proper descendants of [inner] we've visited.
-*)
-let rec compress t ~inner_node ~inner ~descendants =
-  match t.node with
-  | Root r ->
-    (* t is the root of the tree.
-       Re-point all descendants directly to it by setting them to [Inner t].
-       Note: we don't re-point [inner] as it already points there. *)
-    List.iter descendants ~f:(fun t -> t.node <- inner_node); (t, r)
-  | Inner t' as node ->
-    compress t' ~inner_node:node ~inner:t ~descendants:(inner :: descendants)
+(* NOTE: does not use tail-recursion like previous implementation, because
+   the depth should never exceed O(log N) anyway.  It's faster this way. *)
+let rec compress ~repr:(Inner inner as repr) = function
+  | Node (Root _ as root) -> repr, root
+  | Node (Inner next_inner as repr) ->
+      let repr, _ as res = compress ~repr next_inner.parent in
+      inner.parent <- Node repr;
+      res
 
-let representative t =
-  match t.node with
-  | Root r -> (t, r)
-  | Inner t' as node -> compress t' ~inner_node:node ~inner:t ~descendants:[]
+let compress_inner (Inner inner as repr) = compress ~repr inner.parent
 
-let root t = snd (representative t)
```

Comment:
You removed `root` without changing all its use sites (eg. all the ones in test module are left untouched).